simplex/engine.go

```diff
@@ -0,0 +1,324 @@
+	blockBuilder := &BlockBuilder{
+		vm:           config.VM,
+		blockTracker: blockTracker,
+		log:          config.Log,
+	}
+	storage.blockTracker = blockTracker
+
+	blockDeserializer := &blockDeserializer{
+		parser:       config.VM,
+		blockTracker: blockTracker,
+	}
+
+	epochConfig := simplex.EpochConfig{
+		MaxProposalWait:     config.Params.MaxProposalWait,
+		MaxRebroadcastWait:  config.Params.MaxRebroadcastWait,
+		QCDeserializer:      qcDeserializer,
+		Logger:              config.Log,
+		ID:                  config.Ctx.NodeID[:],
+		Signer:              &signer,
+		Verifier:            &verifier,
+		BlockDeserializer:   blockDeserializer,
+		SignatureAggregator: signatureAggregator,
+		Comm:                comm,
+		Storage:             storage,
+		WAL:                 config.WAL,
+		BlockBuilder:        blockBuilder,
+		Epoch:               simplexBlock.metadata.Epoch,
+		StartTime:           time.Now(),
+		ReplicationEnabled:  true,
+	}
+
+	epoch, err := simplex.NewEpoch(epochConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Engine{
+		epoch:                       epoch,
+		blockDeserializer:           blockDeserializer,
+		quorumDeserializer:          qcDeserializer,
+		logger:                      config.Log,
+		StateSummaryFrontierHandler: common.NewNoOpStateSummaryFrontierHandler(config.Log),
+		AcceptedStateSummaryHandler: common.NewNoOpAcceptedStateSummaryHandler(config.Log),
+		AcceptedFrontierHandler:     common.NewNoOpAcceptedFrontierHandler(config.Log),
+		AcceptedHandler:             common.NewNoOpAcceptedHandler(config.Log),
+		AncestorsHandler:            common.NewNoOpAncestorsHandler(config.Log),
+		PutHandler:                  common.NewNoOpPutHandler(config.Log),
+		QueryHandler:                common.NewNoOpQueryHandler(config.Log),
+		ChitsHandler:                common.NewNoOpChitsHandler(config.Log),
+		Connector:                   config.VM,
+
+		tickInterval: getTickInterval(config.Params),
+		AppHandler:   config.VM,
+		vm:           config.VM,
+		consensusCtx: consensusCtx,
+	}, nil
+}
+
+func (e *Engine) Start(ctx context.Context, _ uint32) error {
+	if e.nonValidator {
+		e.logger.Info("non-validator cannot start simplex engine")
+		return nil
+	}
+
+	e.logger.Info("Starting simplex engine")
+	go e.tick()
+	err := e.epoch.Start()
+	if err != nil {
+		return fmt.Errorf("failed to start simplex epoch: %w", err)
+	}
+
+	e.consensusCtx.State.Set(snow.EngineState{
+		Type:  p2p.EngineType_ENGINE_TYPE_CHAIN,
+		State: snow.NormalOp,
+	})
+	return e.vm.SetState(ctx, snow.NormalOp)
+}
+
+func getTickInterval(params *pSimplex.Parameters) time.Duration {
+	tick := min(int64(params.MaxProposalWait), int64(params.MaxRebroadcastWait)) / 10
+	return time.Duration(tick)
+}
+
+// tick periodically advances the engine's time.
+func (e *Engine) tick() {
+	ticker := time.NewTicker(e.tickInterval)
+	for {
+		if e.shutdown {
+			return
+		}
+
+		tick := <-ticker.C
```

Comment:
let's add a close channel to the struct so we can close shutdown from multiple goroutines. 

Then, we don't need `e.shutdown` anymore, and we can immediately return here via:

```
func (e *Engine) tick() {
	ticker := time.NewTicker(e.tickInterval)
	for {
		select {
		case tick := <-ticker.C:
			e.epoch.AdvanceTime(tick)
		case <-e.shutdown:
			ticker.Stop()
			return
		}
	}
}
```